Sources/SmithyCBOR/Reader/Reader.swift

```diff
@@ -0,0 +1,237 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+import Foundation
+
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyReader
+@_spi(SmithyReadWrite) import enum SmithyReadWrite.ReaderError
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyWriter
+@_spi(Smithy) import struct Smithy.Document
+@_spi(Smithy) import protocol Smithy.SmithyDocument
+@_spi(SmithyTimestamps) import enum SmithyTimestamps.TimestampFormat
+@_spi(SmithyTimestamps) import struct SmithyTimestamps.TimestampFormatter
+
+@_spi(SmithyReadWrite)
+public final class Reader: SmithyReader {
+    public typealias NodeInfo = String
+
+    public let cborValue: CBORType?
+    public let nodeInfo: NodeInfo
+    public internal(set) var children: [Reader] = []
+    public internal(set) weak var parent: Reader?
+    public var hasContent: Bool { cborValue != nil && cborValue != .null }
+
+    public init(nodeInfo: NodeInfo, cborValue: CBORType?, parent: Reader? = nil) {
+        self.nodeInfo = nodeInfo
+        self.cborValue = cborValue
+        self.parent = parent
+        self.children = Self.children(from: cborValue, parent: self)
+    }
+
+    public static func from(data: Data) throws -> Reader {
```

Comment:
this function calls CBORDecoder which is defined in CRT for actually decoding individual types